assembl/graphql/extract.py

```diff
@@ -0,0 +1,111 @@
+import graphene
+from graphene.relay import Node
+from graphene_sqlalchemy import SQLAlchemyObjectType
+
+from pyramid.httpexceptions import HTTPUnauthorized
+from pyramid.security import Everyone
+
+from assembl.auth import CrudPermissions
+from assembl.auth.util import get_permissions
+from assembl import models
+
+from .types import SecureObjectType, SQLAlchemyInterface
+from .utils import abort_transaction_on_exception, DateTime
+from .user import AgentProfile
+
+
+class TextFragmentIdentifier(SecureObjectType, SQLAlchemyObjectType):
+    class Meta:
+        model = models.TextFragmentIdentifier
+        interfaces = (Node,)
+        only_fields = ('xpath_start', 'xpath_end', 'offset_start', 'offset_end')
+
+
+class ExtractInterface(SQLAlchemyInterface):
+    class Meta:
+        model = models.Extract
+        only_fields = ('body', 'important', 'extract_nature', 'extract_action')
+        # Don't add id in only_fields in an interface or the the id of Post
+        # will be just the primary key, not the base64 type:id
+
+    text_fragment_identifiers = graphene.List(TextFragmentIdentifier)
+
+
+class Extract(SecureObjectType, SQLAlchemyObjectType):
+    class Meta:
+        model = models.Extract
+        interfaces = (Node, ExtractInterface)
+        only_fields = ('id')
+
+    creation_date = DateTime()
+    creator_id = graphene.Int()
+    creator = graphene.Field(lambda: AgentProfile)
+
+    def resolve_creator(self, args, context, info):
+        if self.creator_id is not None:
+            return models.AgentProfile.get(self.creator_id)
+
+
+class UpdateExtract(graphene.Mutation):
+    class Input:
+        extract_id = graphene.ID(required=True)
+        idea_id = graphene.ID()
+        important = graphene.Boolean()
+        extract_nature = graphene.String()
+        extract_action = graphene.String()
+
+    extract = graphene.Field(lambda: Extract)
+
+    @staticmethod
+    @abort_transaction_on_exception
+    def mutate(root, args, context, info):
+        discussion_id = context.matchdict['discussion_id']
+
+        user_id = context.authenticated_userid or Everyone
+
+        permissions = get_permissions(user_id, discussion_id)
+        allowed = models.Extract.user_can_cls(
+            user_id, CrudPermissions.UPDATE, permissions)
+        if not allowed:
+            raise HTTPUnauthorized()
+
+        extract_id = args.get('extract_id')
+        extract_id = int(Node.from_global_id(extract_id)[1])
+        extract = models.Extract.get(extract_id)
+        extract.important = args.get('important')
```

Comment:
If import is optional here, you should maybe fallback to False instead of None explictly?